docs/proposals/hpa/cronfederatedhpa.md

```diff
@@ -0,0 +1,339 @@
+
+### Components change
+
+#### karmada-controller-manager  
+In order to make `CronFederatedHPA` work, CronFederatedHPA controller should be implemented in `karmada-controller-manager`, it will check the time and apply the rules of `CronFederatedHPA` at the configured time. The timestamp is configured via `spec.rules[*].schedule`, which should restrict the following format:
+```sh
+# ┌───────────── minute (0 - 59)
+# │ ┌───────────── hour (0 - 23)
+# │ │ ┌───────────── day of the month (1 - 31)
+# │ │ │ ┌───────────── month (1 - 12)
+# │ │ │ │ ┌───────────── day of the week (0 - 6) (Sunday to Saturday;
+# │ │ │ │ │                                   7 is also Sunday on some systems)
+# │ │ │ │ │                                   OR sun, mon, tue, wed, thu, fri, sat
+# │ │ │ │ │
+# * * * * *
+```
+For example:
+| Description																									| Equivalent to |
+| ------------- 																							|-------------  |
+| Run once a year at midnight of 1 January										| 0 0 1 1 * 		|
+| Run once a month at midnight of the first day of the month	| 0 0 1 * * 		|
+| Run once a week at midnight on Sunday morning								| 0 0 * * 0 		|
+| Run once a day at midnight																	| 0 0 * * * 		|
+| Run once an hour at the beginning of the hour								| 0 * * * * 		|
+
+The process of rule state(showed in the `status` field) transition is as follows:  
+<image src="./statics/rule-state-transition.png" width="600px">
+
+
+To better understand what is next schedule time, here are some examples:
+* The schedule time is `3 * * * *`, I apply this rule when 9:04 AM, so the next schedule time is: 10:03 AM, after executing this rule, the next time will be updated to 11:03 AM.
+* The schedule time is `3 * * * *`, I apply this rule when 9:01 AM, so the next schedule time is: 09:03 AM, after executing this rule, the next time will be updated to 10:03 AM.
+
+#### karmada-webhook
+In order to make sure the applied configuration is corrent, some validations are necessary for `CronFederatedHPA`, these logic should be implemented in `karmada-webhook`:
+* If `spec.scaleTargetRef.apiVersion` is `autoscaling.karmada.io/v1alpha1` and `spec.scaleTargetRef.kind` is `FederatedHPA`, `spec.rules[*].targetMinReplicas` and `spec.rules[*].targetMaxReplicas` cannot be empty at the same time.
+* If `spec.scaleTargetRef.kind` is not `autoscaling.karmada.io/v1alpha1`, `spec.rules[*].targetReplicas` cannot be empty.
+* `spec.rules[*].schedule` should be a valid cron format.
+* `maxDelaySeconds` should be smaller than the period interval.
+
+### Story Solution
+
+#### Story 1
+
+Requirements:
+* Scale up workloads to 1000 every day at 08:30 AM.
+* Scale down workloads to 1 every day at 11:00 AM.
+
+
+```yaml
+apiVersion: autoscaling.karmada.io/v1alpha1
+kind: CronFederatedHPA
+metadata:
+  name: cron-federated-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: app/v1
+    kind: Deployment
+    name: shop
+  rules:
```

Comment:
I would like to provide some warnings here. Cron has the functionality to self-integrate both the beginning time and the end time and whatever combination you can think of. Further, the https://pkg.go.dev/github.com/robfig/cron itself seems to follow these rules strictly. This same feature has caused a similar bug in KEDA as they have both the start time schedule and the end time schedule. See
triggers:

https://github.com/kedacore/keda/issues/4677
  triggers:
  - metadata:
      desiredReplicas: "3"
      end: 10 22 * * *
      start: 30 6 * * 1-5
      timezone: Europe/Prague

We should either just go with a single cron schedule and two field scaleUpTargetReplicas and scaleDownTargetReplica. 

There is one more design as well that the KEDA community came up with after looking at the kube-green way. I think this one fits better for the KEDA use case as they use Adapter in between to connect their scalers with the rest of the design. But here as we are directly aiming to connect with Kubernetes API using cron object directly would be a better decision as Kubernetes can handle them natively and doing it with the below-given design, would add extra parsing. https://github.com/kedacore/keda/issues/3566 here is the given design.  
- type: time-window
  weekdays: "Mo-Fri"
  start: "08:00"
  end: "20:00"
  timezone: Asia/Kolkata  # The acceptable values would be a value from the IANA Time Zone Database.

Reply:
I check the issue, the problem exists because of the implementation way in KEDA(https://github.com/kedacore/keda/blob/2234a6f30df6b9fa83a1b88f1aa609ebee846983/pkg/scalers/cron_scaler.go#L164-L174)
Here we do not define the end time with a schedule time format. So I think we will not have this problem.

Reply:
But in any case we would have to add a behavior to handles cases such as 30 6 * * 1-5 where or something else. In any case if user gives us both the upper and lower limit we would have to handle the scenario than. And if we can take time for both using a single parameter it would increase our code redundancy. But I am not sure of the exact implementation that we are sticking to so if we have something else in our mind, during implementation, then we can go ahead with the above approach but in my opinion, if the aim of those parameters is just to act as start and end schedule resembler, then it would ultimately lead to extra parsing and rules limiting effort to avoid bugs and it would ultimately increase redundant code as well.